### PR TITLE
Fix Hamiltonian application for nonstandard models

### PR DESCRIPTION
**Pull request by antoine-levitt**
Has been broken since https://github.com/JuliaMolSim/DFTK.jl/commit/530548459ce95dd6bb9dceb0748f7c3838b52eba (and was only accidentally working before). Good thing nobody uses magnetic fields, huh :-)

### CODE REVIEW

File: test/anyons.jl [hunk 1]
```diff
@@ -18,3 +18,35 @@ using Test
     @test norm(curlA - 2π*DFTK.ρref_real_2D(x, y, M, σ)) < 1e-4
     @test abs(divA) < 1e-6
 end
+
+@testset "Anyons: check E11" begin
```
Comment (mfherbst) [inline test/anyons.jl:22]: What is E11 ?

File: src/terms/Hamiltonian.jl [hunk 2]
```diff
@@ -229,5 +231,5 @@ end
 function hamiltonian_with_total_potential(Hk::HamiltonianBlock, V)
     operators = [op for op in Hk.operators if !(op isa RealSpaceMultiplication)]
     push!(operators, RealSpaceMultiplication(Hk.basis, Hk.kpoint, V))
-    HamiltonianBlock(Hk.basis, Hk.kpoint, operators, Hk.scratch)
+    HamiltonianBlock(Hk.basis, Hk.kpoint, operators)
```
Comment (mfherbst) [inline src/terms/Hamiltonian.jl:234]: Why remove this? Saves quite a bit of allocation for potential-based SCFs

File: test/anyons.jl (hunk 1 above)
Comment (antoine-levitt) [reply test/anyons.jl:22]: Some quantity defined in the paper, doesn't really matter what it is, it's just a test we reproduce the results in the paper

File: src/terms/Hamiltonian.jl (hunk 2 above)
Comment (antoine-levitt) [reply src/terms/Hamiltonian.jl]: OK, I added it back. It's a bit ugly, but this is a pretty internal thing anyway

**Pull request merged**